### PR TITLE
Fix for Missing or empty metadata values are not allowed with POST /v2/op/update #3121

### PR DESCRIPTION
**Pull request by SourabhChourasiya-NEC**
Hi @fgalan Sir,
I have worked on this issue and raising the PR for the same.
As you referenced the document, in which it is clearly mentioned that `Attribute/metadata value may be omitted in requests, meaning that the attribute/metadata has nullvalue. In responses, the value is always present.`
Hence, this fix will work for only NGSIv2 APIs.

I also want to inform you that for `Empty string `, request and response is same as mentioned in the issue #3121. but for `Missing value`,  I'm getting the different response before applying the fix.

For Empty string
```
POST /v2/op/update

{
  "actionType": "APPEND",
  "entities": [
    {
      "type": "testing",
      "id": "e1",
      "a": {
      	"type": "Text",
      	"value": "foo",
      	"metadata": {
      		"name": {
      			"type": "T",
      			"value": ""
      		}
      	}
      }
    }
  ]
}

400 Bad  Request
{
    "error": "BadRequest",
    "description": "missing metadata value"
}
```
For Missing value
```
POST /v2/op/update

{
  "actionType": "APPEND",
  "entities": [
    {
      "type": "testing",
      "id": "e1",
      "a": {
      	"type": "Text",
      	"value": "foo",
      	"metadata": {
      		"name": {
      			"type": "T"
      		}
      	}
      }
    }
  ]
}

204 No Content
```

Because check is only available for empty string in code for metadata value `(src/lib/ngsi/Metadata.cpp)`. 
I've also performed functional test & unit test successfully.

Thanks

### CODE REVIEW

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
Please guide me further so we can close this issue.
Thank you

**Comment by fgalan** [comment]
First of all, thanks for your contribution! :)

Next step would be to create a .test in order to ensure that the fix works (and will not break in the future with new modifications). That .test file (or serveral .test fiels, if using just one is too complex) should include different ways of creating missing and empty value metadata, in particular:

* POST /v2/op/update
* POST /v2/entities
* POST /v2/entities/<id>/attrs (adding a new attribute with missing and empty value metadata)
* PUT /v2/entities/<id>/attrs/<attrName> (updating and existing attribute with missing and empty value metadata)
* (If you find some request I'm missing, please add it also)

and the corresponding GET operations to check that everything is as expected.

In addition, a line describing the fix in the CHANGE_NEXT_RELEASE file should be included.

**Comment by fgalan** [comment]
You can find information about how to create a .test in the following link: https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/devel/cookbook.md#adding-a-functional-test-case

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
I have added all the FT cases covering all the create and update request for NGSIv2 as suggested by you.
Please review it and guide me further.
Thank You

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -2,3 +2,4 @@
 - Hardening: Mongo driver now compiled using --use-sasl-client --ssl to enable proper DB authentication mechanisms
 - Fix: correct error payload using errorCode (previously orionError was used) in POST /v1/queryContext and POST /v1/updateContext in some cases
 - Hardening: modification of the URL parsing mechanism, making it more efficient, and the source code easier to follow (#3109, step 1)
+- Fix : Fix for Missing or empty metadata values are not allowed with POST /v2/op/update (#3121)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: I'd suggest the following wording, which I think is more precise:

```
- Fix : Missing or empty metadata values were not allowed in NGSIv1 create/update operations (#3121)
```

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Missing or empty metadata values are not allowed with POST /v2/op/update
```
Comment (fgalan) [inline test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: In this case, the test is not about POST /v2/op/update, so I should be:

> Missing or empty metadata values are not allowed with POST /v2/entities/id/attrs

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
```
Comment (fgalan) [inline test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: It use to happend provide some extra info in the test script (and echo lines that are based on it). For instance:

> # 01. POST /v2/entities, to create Room1 with temperature (no metadata)
> # 02. GET /v2/entities/Room1
> # 03. POST /v2/entities/Room1/attrs, to add metadata with "" value to temperature and add pressure attribute with metadata with no value
> # 04. GET /v2/entities/Room1
> # 05. POST /v2/entities/Room1/attrs, to update temperature with metadata with no value, pressure with metadata with regular value and add humidity attribute with metadata with "" value
> # 06. GET /v2/entities/Room1

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (SourabhChourasiya-NEC) [reply CHANGES_NEXT_RELEASE]: fixed in 991cd65

File: CHANGES_NEXT_RELEASE [hunk 4]
```diff
@@ -2,3 +2,4 @@
 - Hardening: Mongo driver now compiled using --use-sasl-client --ssl to enable proper DB authentication mechanisms
 - Fix: correct error payload using errorCode (previously orionError was used) in POST /v1/queryContext and POST /v1/updateContext in some cases
 - Hardening: modification of the URL parsing mechanism, making it more efficient, and the source code easier to follow (#3109, step 1)
+- Fix : Missing or empty metadata values were not allowed in NGSIv1 create/update operations (#3121)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: NGSIv1 -> NGSIv2

My fault... I told NGSIv1 firs time (it was a typo) :(

Comment (SourabhChourasiya-NEC) [reply CHANGES_NEXT_RELEASE]: fixed in 25fa92f

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test (hunk 3 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in ea851ec

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test (hunk 2 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in ea851ec

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
I have updated the Ft Case file as suggested by you.
Please review it and guide me further.

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+# 01. POST /v2/entities, to create Room1 with temperature (no metadata)
+# 02. GET /v2/entities/Room1
+# 03. POST /v2/entities/Room1/attrs, to add metadata with "" value to temperature and add pressure attribute with metadata with no value
+# 04. GET /v2/entities/Room1
+# 05. POST /v2/entities/Room1/attrs, to update temperature with metadata with no value, pressure with metadata with regular value and add humidity attribute with metadata with "" value
+# 06. GET /v2/entities/Room1
+# 07. POST /v2/entities, to create Room2 with temperature having two metadata one with "" value and another without value
+# 08. GET /v2/entities/Room2
+# 09. PUT /v2/entities/Room2/attrs/temperature/value , updating the value of Room2 temperature attribute.
```
Comment (fgalan) [inline test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: New description are good! :)

However, please include them also in the echo lines for each step. I mean change for instance:

```
 echo "03. POST /v2/entities/Room1/attrs." 
```
to 
```
echo '03. POST /v2/entities/Room1/attrs, to add metadata with "" value to temperature and add pressure attribute with metadata with no value'
```

File: CHANGES_NEXT_RELEASE [hunk 6]
```diff
@@ -2,3 +2,4 @@
 - Hardening: Mongo driver now compiled using --use-sasl-client --ssl to enable proper DB authentication mechanisms
 - Fix: correct error payload using errorCode (previously orionError was used) in POST /v1/queryContext and POST /v1/updateContext in some cases
 - Hardening: modification of the URL parsing mechanism, making it more efficient, and the source code easier to follow (#3109, step 1)
+- Fix : Missing or empty metadata values were not allowed in NGSIv2 create/update operations (#3121)
```
[the remark this answers is not in the document]
Comment (SourabhChourasiya-NEC) [reply CHANGES_NEXT_RELEASE]: Fixed in 18091e5

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test (hunk 5 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in dc6cf02

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+echo "==========================="
+orionCurl --url /v2/entities/Room2
+echo
+echo
+
+
+echo "09. PUT /v2/entities/Room2/attrs/temperature/value."
+echo "==================================================="
+payload=' 25.6 '
+orionCurl --url /v2/entities/Room2/attrs/temperature/value --payload "$payload" --in text/plain --out json -X PUT
```
[the remark this answers is not in the document]
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in dc6cf02

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Missing or empty metadata values are not allowed with POST /v2/entities/id/attrs
```
[the remark this answers is not in the document]
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in dc6cf02

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
I have made all the necessary changes in FT case file as guided by you.
Please review it and guide me further
Thanks

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+echo "08. GET /v2/entities/Room2."
+echo "==========================="
+orionCurl --url /v2/entities/Room2
+echo
+echo
+
+
+echo "09. PUT /v2/entities/Room2/attrs/temperature/value, updating the value of Room2 temperature attribute."
+echo "======================================================================================================"
+payload='{
```
Comment (fgalan) [inline test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: Note the accuracy and tolerance metadata values are the same used in step 7. They should be different (e.g. the oposite, i.e. accuracy with missing value and tolerance with "" value) in order to check that step 9 is actually working properly.

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+            }
+        },
+        "type": "Float",
+        "value": 25
+    },
+    "type": "Room"
+}
+
+
+05. POST /v2/entities/Room1/attrs, to update temperature with metadata with no value, pressure with metadata with regular value and add humidity attribute with metadata with  value.
```
Comment (fgalan) [inline test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: Better use `echo '...'` when you have `"` in the message you want to print (as in this case and some others in the tests). In addition, also as general comment, try that the line with `=` have the same length that the message printed above.

**Comment by fgalan** [comment]
I have posted a couple of final minor comments. As soon as they are addressed, I think we are done :)

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test (hunk 9 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in bcf8353

File: test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test (hunk 10 above)
Comment (SourabhChourasiya-NEC) [reply test/functionalTest/cases/3121_missing_or_empty_metadata_values/entitiy_creation.test]: fixed in bcf8353

**Pull request merged**